File: packages/mobile-android/sdk/src/main/java/com/rtcexpress/sdk/P2pMediaEngine.kt
package com.rtcexpress.sdk

import android.content.Context
import org.json.JSONObject
import org.webrtc.AudioSource
import org.webrtc.AudioTrack
import org.webrtc.Camera2Enumerator
import org.webrtc.DefaultVideoDecoderFactory
import org.webrtc.DefaultVideoEncoderFactory
import org.webrtc.EglBase
import org.webrtc.IceCandidate
import org.webrtc.MediaConstraints
import org.webrtc.PeerConnection
import org.webrtc.PeerConnectionFactory
import org.webrtc.SessionDescription
import org.webrtc.SurfaceTextureHelper
import org.webrtc.SurfaceViewRenderer
import org.webrtc.VideoCapturer
import org.webrtc.VideoSource
import org.webrtc.VideoTrack

class P2pMediaEngine(
    private val context: Context,
    private val userId: String,
    private val sendSignaling: (String, JSONObject) -> Unit,
    private val activeCall: () -> Pair<String, String>?
) {
    private val eglBase = EglBase.create()
    private val factory: PeerConnectionFactory
    private var peerConnection: PeerConnection? = null
    private var audioSource: AudioSource? = null
    private var audioTrack: AudioTrack? = null
    private var videoSource: VideoSource? = null
    private var videoTrack: VideoTrack? = null
    private var videoCapturer: VideoCapturer? = null
    private var video = false
    private var usingFrontCamera = true

    var onRemoteVideoTrack: ((VideoTrack) -> Unit)? = null

    init {
        PeerConnectionFactory.initialize(
            PeerConnectionFactory.InitializationOptions.builder(context)
                .createInitializationOptions()
        )
        factory = PeerConnectionFactory.builder()
            .setVideoEncoderFactory(DefaultVideoEncoderFactory(eglBase.eglBaseContext, true, true))
            .setVideoDecoderFactory(DefaultVideoDecoderFactory(eglBase.eglBaseContext))
            .createPeerConnectionFactory()
    }

    fun prepare(enableVideo: Boolean) {
        video = enableVideo
        if (peerConnection != null) return

        val iceServers = listOf(
            PeerConnection.IceServer.builder("stun:stun.l.google.com:19302").createIceServer()
        )
        peerConnection = factory.createPeerConnection(
            iceServers,
            object : PeerConnection.Observer {
                override fun onIceCandidate(candidate: IceCandidate) {
                    val call = activeCall() ?: return
                    val payload = JSONObject()
                        .put("callId", call.first)
                        .put("fromUserId", userId)
                        .put("toUserId", call.second)
                        .put(
                            "candidate",
                            JSONObject()
                                .put("candidate", candidate.sdp)
                                .put("sdpMid", candidate.sdpMid)
                                .put("sdpMLineIndex", candidate.sdpMLineIndex)
                        )
                    sendSignaling("ice_candidate", payload)
                }

                override fun onAddTrack(receiver: org.webrtc.RtpReceiver?, streams: Array<out org.webrtc.MediaStream>?) {
                    val track = receiver?.track()
                    if (track is VideoTrack) onRemoteVideoTrack?.invoke(track)
                }

                override fun onSignalingChange(state: PeerConnection.SignalingState?) {}
                override fun onIceConnectionChange(state: PeerConnection.IceConnectionState?) {}
                override fun onIceConnectionReceivingChange(receiving: Boolean) {}
                override fun onIceGatheringChange(state: PeerConnection.IceGatheringState?) {}
                override fun onIceCandidatesRemoved(candidates: Array<out IceCandidate>?) {}
                override fun onAddStream(stream: org.webrtc.MediaStream?) {}
                override fun onRemoveStream(stream: org.webrtc.MediaStream?) {}
                override fun onDataChannel(channel: org.webrtc.DataChannel?) {}
                override fun onRenegotiationNeeded() {}
                override fun onRemoveTrack(receiver: org.webrtc.RtpReceiver?) {}
            }
        )

        val audioConstraints = MediaConstraints()
        audioSource = factory.createAudioSource(audioConstraints)
        audioTrack = factory.createAudioTrack("audio0", audioSource)
        peerConnection?.addTrack(audioTrack)

        if (enableVideo) {
            startCamera()
            videoTrack?.let { peerConnection?.addTrack(it) }
        }
    }

    fun createOffer(peerUserId: String, callId: String) {
        prepare(video)
        peerConnection?.createOffer(object : SimpleSdpObserver("offer") {
            override fun onCreateSuccess(description: SessionDescription) {
                peerConnection?.setLocalDescription(SimpleSdpObserver("local-offer"), description)
                val payload = JSONObject()
                    .put("callId", callId)
                    .put("fromUserId", userId)
                    .put("toUserId", peerUserId)
                    .put("sdp", JSONObject().put("type", description.type.canonicalForm()).put("sdp", description.description))
                sendSignaling("webrtc_offer", payload)
            }
        }, MediaConstraints())
    }

    fun handleOffer(payload: JSONObject) {
        prepare(video)
        val sdp = payload.getJSONObject("sdp")
        val remote = SessionDescription(SessionDescription.Type.OFFER, sdp.getString("sdp"))
        peerConnection?.setRemoteDescription(SimpleSdpObserver("remote-offer"), remote)
        peerConnection?.createAnswer(object : SimpleSdpObserver("answer") {
            override fun onCreateSuccess(description: SessionDescription) {
                peerConnection?.setLocalDescription(SimpleSdpObserver("local-answer"), description)
                val answer = JSONObject()
                    .put("callId", payload.getString("callId"))
                    .put("fromUserId", userId)
                    .put("toUserId", payload.getString("fromUserId"))
                    .put("sdp", JSONObject().put("type", description.type.canonicalForm()).put("sdp", description.description))
                sendSignaling("webrtc_answer", answer)
            }
        }, MediaConstraints())
    }

    fun handleAnswer(payload: JSONObject) {
        val sdp = payload.getJSONObject("sdp")
        val remote = SessionDescription(SessionDescription.Type.ANSWER, sdp.getString("sdp"))
        peerConnection?.setRemoteDescription(SimpleSdpObserver("remote-answer"), remote)
    }

    fun handleIceCandidate(payload: JSONObject) {
        val candidate = payload.getJSONObject("candidate")
        val ice = IceCandidate(
            candidate.optString("sdpMid"),
            candidate.optInt("sdpMLineIndex"),
            candidate.getString("candidate")
        )
        peerConnection?.addIceCandidate(ice)
    }

    fun attachLocalVideo(renderer: SurfaceViewRenderer) {
        renderer.init(eglBase.eglBaseContext, null)
        renderer.setMirror(usingFrontCamera)
        videoTrack?.addSink(renderer)
    }

    fun attachRemoteVideo(track: VideoTrack, renderer: SurfaceViewRenderer) {
        renderer.init(eglBase.eglBaseContext, null)
        track.addSink(renderer)
    }

    fun switchCamera() {
        if (!video) return
        (videoCapturer as? org.webrtc.CameraVideoCapturer)?.switchCamera(null)
        usingFrontCamera = !usingFrontCamera
    }

    fun muteMicrophone(muted: Boolean) {
        audioTrack?.setEnabled(!muted)
    }

    fun muteCamera(muted: Boolean) {
        videoTrack?.setEnabled(!muted)
    }

    fun destroy() {
        videoCapturer?.stopCapture()
        videoCapturer?.dispose()
        videoTrack?.dispose()
        videoSource?.dispose()
        audioTrack?.dispose()
        audioSource?.dispose()
        peerConnection?.close()
        peerConnection = null
        factory.dispose()
        eglBase.release()
    }

    private fun startCamera() {
        val enumerator = Camera2Enumerator(context)
        val device = enumerator.deviceNames.firstOrNull { enumerator.isFrontFacing(it) } ?: return
        videoCapturer = enumerator.createCapturer(device, null)
        val helper = SurfaceTextureHelper.create("capture", eglBase.eglBaseContext)
        videoSource = factory.createVideoSource(videoCapturer!!.isScreencast)
        videoCapturer!!.initialize(helper, context, videoSource!!.capturerObserver)
        videoCapturer!!.startCapture(640, 480, 24)
        videoTrack = factory.createVideoTrack("video0", videoSource)
    }

    private open class SimpleSdpObserver(private val label: String) : org.webrtc.SdpObserver {
        override fun onCreateSuccess(description: SessionDescription) {}
        override fun onSetSuccess() {}
        override fun onCreateFailure(error: String) {}
        override fun onSetFailure(error: String) {}
    }
}

File: packages/mobile-ios/Sources/RTCExpress/P2pMediaEngine.swift
import Foundation
import WebRTC

final class P2pMediaEngine: NSObject {
    private let userId: String
    private let sendSignaling: (String, [String: Any]) throws -> Void
    private let activeCall: () -> (callId: String, peerUserId: String)?

    private var factory: RTCPeerConnectionFactory?
    private var peerConnection: RTCPeerConnection?
    private var audioTrack: RTCAudioTrack?
    private var videoTrack: RTCVideoTrack?
    private var capturer: RTCCameraVideoCapturer?
    private var videoEnabled = false
    private var usingFrontCamera = true

    var onRemoteVideoTrack: ((RTCVideoTrack) -> Void)?

    init(
        userId: String,
        sendSignaling: @escaping (String, [String: Any]) throws -> Void,
        activeCall: @escaping () -> (callId: String, peerUserId: String)?
    ) {
        self.userId = userId
        self.sendSignaling = sendSignaling
        self.activeCall = activeCall
        super.init()
        RTCInitializeSSL()
        let encoder = RTCDefaultVideoEncoderFactory()
        let decoder = RTCDefaultVideoDecoderFactory()
        factory = RTCPeerConnectionFactory(encoderFactory: encoder, decoderFactory: decoder)
    }

    func prepare(video: Bool) {
        videoEnabled = video
        guard peerConnection == nil, let factory else { return }

        let config = RTCConfiguration()
        config.iceServers = [RTCIceServer(urlStrings: ["stun:stun.l.google.com:19302"])]
        let constraints = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil)
        peerConnection = factory.peerConnection(with: config, constraints: constraints, delegate: self)

        let audioSource = factory.audioSource(with: constraints)
        audioTrack = factory.audioTrack(with: audioSource, trackId: "audio0")
        peerConnection?.add(audioTrack!, streamIds: ["stream0"])

        if video {
            startCamera(factory: factory)
            if let videoTrack {
                peerConnection?.add(videoTrack, streamIds: ["stream0"])
            }
        }
    }

    func createOffer(peerUserId: String, callId: String) {
        prepare(video: videoEnabled)
        let constraints = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil)
        peerConnection?.offer(for: constraints) { [weak self] sdp, _ in
            guard let self, let sdp else { return }
            self.peerConnection?.setLocalDescription(sdp)
            let payload: [String: Any] = [
                "callId": callId,
                "fromUserId": self.userId,
                "toUserId": peerUserId,
                "sdp": ["type": sdp.type.rawValue, "sdp": sdp.sdp]
            ]
            try? self.sendSignaling("webrtc_offer", payload)
        }
    }

    func handleOffer(_ payload: [String: Any]) {
        prepare(video: videoEnabled)
        guard let sdpObj = payload["sdp"] as? [String: Any],
              let sdpText = sdpObj["sdp"] as? String,
              let type = sdpObj["type"] as? String else { return }
        let remote = RTCSessionDescription(type: RTCSdpType(rawValue: type) ?? .offer, sdp: sdpText)
        peerConnection?.setRemoteDescription(remote) { [weak self] _ in
            guard let self else { return }
            let constraints = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil)
            self.peerConnection?.answer(for: constraints) { answer, _ in
                guard let answer else { return }
                self.peerConnection?.setLocalDescription(answer)
                let response: [String: Any] = [
                    "callId": payload["callId"] as? String ?? "",
                    "fromUserId": self.userId,
                    "toUserId": payload["fromUserId"] as? String ?? "",
                    "sdp": ["type": answer.type.rawValue, "sdp": answer.sdp]
                ]
                try? self.sendSignaling("webrtc_answer", response)
            }
        }
    }

    func handleAnswer(_ payload: [String: Any]) {
        guard let sdpObj = payload["sdp"] as? [String: Any],
              let sdpText = sdpObj["sdp"] as? String,
              let type = sdpObj["type"] as? String else { return }
        let remote = RTCSessionDescription(type: RTCSdpType(rawValue: type) ?? .answer, sdp: sdpText)
        peerConnection?.setRemoteDescription(remote)
    }

    func handleIceCandidate(_ payload: [String: Any]) {
        guard let candidateObj = payload["candidate"] as? [String: Any],
              let candidate = candidateObj["candidate"] as? String else { return }
        let ice = RTCIceCandidate(
            sdp: candidate,
            sdpMLineIndex: Int32(candidateObj["sdpMLineIndex"] as? Int ?? 0),
            sdpMid: candidateObj["sdpMid"] as? String
        )
        peerConnection?.add(ice)
    }

    func attachLocalVideo(to renderer: RTCVideoRenderer) {
        videoTrack?.add(renderer)
    }

    func attachRemoteVideo(_ track: RTCVideoTrack, to renderer: RTCVideoRenderer) {
        track.add(renderer)
    }

    func switchCamera() {
        guard let capturer else { return }
        usingFrontCamera.toggle()
        if let device = RTCCameraVideoCapturer.captureDevices().first(where: {
            $0.position == (usingFrontCamera ? .front : .back)
        }) {
            capturer.startCapture(with: device, format: RTCCameraVideoCapturer.supportedFormats(for: device).last!, fps: 24)
        }
    }

    func muteMicrophone(_ muted: Bool) {
        audioTrack?.isEnabled = !muted
    }

    func muteCamera(_ muted: Bool) {
        videoTrack?.isEnabled = !muted
    }

    func destroy() {
        capturer?.stopCapture()
        peerConnection?.close()
        peerConnection = nil
        factory = nil
        RTCCleanupSSL()
    }

    private func startCamera(factory: RTCPeerConnectionFactory) {
        let source = factory.videoSource()
        capturer = RTCCameraVideoCapturer(delegate: source)
        videoTrack = factory.videoTrack(with: source, trackId: "video0")
        if let device = RTCCameraVideoCapturer.captureDevices().first(where: { $0.position == .front }),
           let format = RTCCameraVideoCapturer.supportedFormats(for: device).last {
            capturer?.startCapture(with: device, format: format, fps: 24)
        }
    }
}

extension P2pMediaEngine: RTCPeerConnectionDelegate {
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {
        if let track = stream.videoTracks.first {
            onRemoteVideoTrack?(track)
        }
    }
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {}
    func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceConnectionState) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceGatheringState) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {
        guard let call = activeCall() else { return }
        let payload: [String: Any] = [
            "callId": call.callId,
            "fromUserId": userId,
            "toUserId": call.peerUserId,
            "candidate": [
                "candidate": candidate.sdp,
                "sdpMid": candidate.sdpMid ?? "",
                "sdpMLineIndex": candidate.sdpMLineIndex
            ]
        ]
        try? sendSignaling("ice_candidate", payload)
    }
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {}
}
